### ISSUE TITLE
Extract region features at a threshold

### ISSUE THREAD

**Issue by kevkid** [facebookresearch/detectron2#1611]
### ❓What I am trying to do with detectron2
I am trying to extract region features where class detection is higher than some threshold. I will be using these features later in my pipeline So far I have trained a Mask R-CNN with this [config](https://github.com/facebookresearch/detectron2/blob/master/configs/COCO-InstanceSegmentation/mask_rcnn_R_101_FPN_3x.yaml) and finetuned it on some custom data. It performs well. What I would like to do is Extract the features from my trained model for the produced bounding box.

### I am trying to do two things:
1. Extract features from bounding boxes
2. Extract features from proposal boxes at a threshold

### 1)
Following along with https://github.com/facebookresearch/detectron2/issues/5 If I am not mistaken I would do the following to extract the features of the **predicted** bounding boxes at a 0.7 threshold (I assume, if the confidence is >0.7 it will produce that instance):

```
#load images
lst = []
for im in outputs_df[outputs_df['is_detected'] == 1]['path'].iloc[:10]:
    img = cv2.imread(im)
    img = cv2.resize(img,(800,800))
    img = np.transpose(img,(2,0,1))
    img_tensor = torch.from_numpy(img)
    lst.append(img_tensor)
images = ImageList.from_tensors(lst[:1], size_divisibility=32).to("cuda")  # preprocessed input tensor
#setup config
cfg = get_cfg()
cfg.merge_from_file(model_zoo.get_config_file("COCO-InstanceSegmentation/mask_rcnn_R_101_FPN_3x.yaml"))
cfg.MODEL.WEIGHTS = os.path.join(cfg.OUTPUT_DIR, "model_final.pth")
cfg.SOLVER.IMS_PER_BATCH = 1
cfg.MODEL.ROI_HEADS.NUM_CLASSES = 1  # only has one class (pnumonia)
#Just run these lines if you have the trained model im memory
cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.7   # set the testing threshold for this model
#build model
model = build_model(cfg)
DetectionCheckpointer(model).load("output/model_final.pth")
model.eval()#make sure its in eval mode

#run model
with torch.no_grad():
    features = model.backbone(images.tensor.float())
    proposals, _ = model.proposal_generator(images, features)
    instances = model.roi_heads._forward_box(features, proposals)
    mask_features = [features[f] for f in model.roi_heads.in_features]
    mask_features = model.roi_heads.mask_pooler(mask_features, [x.pred_boxes for x in instances])
```
Then to extract the features for the bounding boxes:

```
pred_boxes = [x.pred_boxes for x in instances]
rois = model.roi_heads.box_pooler([features[f] for f in model.roi_heads.in_features], pred_boxes)
```
Where `rois` are the region of interest features

- **Does this look right?**
- **How would I do it for proposals (where I can get many more proposed boxes and choose the highest ones)?**

**Comment by ppwwyyxx** [comment]
1. looks right. Although `mask_features` is computed and not used for anything.

2. replace `pred_boxes` by a different object. Output of proposal_generator is documented at https://github.com/facebookresearch/detectron2/blob/ae931b4722d5e6c12464d9c4427545ac7456e0ce/detectron2/modeling/proposal_generator/rpn.py#L388-L390 which we'll add to online documentation soon.

**Comment by kevkid** [comment]
@ppwwyyxx Thank you for letting me know I am on the right track!

### Now onto the second part
I am thinking I have to do something similar:
```
proposal_boxes = [x.proposal_boxes for x in proposals]
proposal_rois = model.roi_heads.box_pooler([features[f] for f in model.roi_heads.in_features], proposal_boxes)
```
This should give me my proposal roi features. This is what I want. But I wanted to know, is there a way for me to figure out which proposals are the the best? Like top 5 most confident proposals, wherein I can use those features of the top most confident proposal boxes as features. 

### Here is what I have so far
I believe I am missing one part and I cant seem to figure out what it is:
```
#found here: https://detectron2.readthedocs.io/_modules/detectron2/modeling/roi_heads/roi_heads.html
box_features = model.roi_heads.box_head(proposal_rois)
predictions = model.roi_heads.box_predictor(box_features)
pred_instances, losses = model.roi_heads.box_predictor.inference(predictions, proposals)
```

If I am not mistaken `predictions` contains: cls_score and bbox_pred (proposal-to-detection box regression deltas) I assume the former has out_features=2 and latter has out_features=4. I also assume we can somehow use this cls_score to figure out which of the proposed boxes have the highest confidence? I tried argmax and it gave me mostly 1's (my classifier has only one class)

Here is an example: 

```
(tensor([[ 0.6308, -0.4926],
         [-1.6662,  1.5430],
         [-0.2080,  0.4856],
         ...,
         [-6.9698,  6.6695],
         [-5.6361,  5.4046],
         [-4.4918,  4.3899]], device='cuda:0', grad_fn=<AddmmBackward>),
```
Thank you very much for your help, I am new to object detection.

**Comment by ppwwyyxx** [comment]
> I assume the former has out_features=2 and latter has out_features=4

Not sure what it means.

> I also assume we can somehow use this cls_score to figure out which of the proposed boxes have the highest confidence

that sounds like one way to go.

> I tried argmax and it gave me mostly 1's

I don't know what/how exactly you tried so I cannot help. The output of `box_predictor` is documented at https://github.com/facebookresearch/detectron2/blob/ae931b4722d5e6c12464d9c4427545ac7456e0ce/detectron2/modeling/roi_heads/fast_rcnn.py#L429-L434 which should have enough information.

**Comment by kevkid** [comment]
>> I tried argmax and it gave me mostly 1's

> I don't know what/how exactly you tried so I cannot help.

Sorry, let me clarify. I don't know how the classes are represented? They aren't probabilities, and I don't know why there are two columns as I only have one class and set `cfg.MODEL.ROI_HEADS.NUM_CLASSES = 1`

Here is how I did it: `predictions[0].softmax(-1).argmax(-1)`

**Comment by ppwwyyxx** [comment]
They are just "scores". The higher the better.
There are K+1=2 classes. The last column is background class.
